test/e2e/specs/site-editor/pages.spec.js

```diff
@@ -18,244 +18,244 @@ async function draftNewPage( page ) {
+	await page
+		.getByRole( 'region', { name: 'Editor top bar' } )
+		.getByRole( 'button', { name: 'Back' } )
+		.click();
 
-		await expect( paragraphsInGroup.nth( 0 ) ).toBeVisible();
-		await expect( paragraphsInGroup.nth( 1 ) ).toBeVisible();
-		// Check order of paragraphs.
-		// Important to ensure the blocks are rendered as they are in the template.
-		await expect( paragraphsInGroup.nth( 0 ) ).toHaveText(
-			'Sweet paragraph 1'
-		);
-		await expect( paragraphsInGroup.nth( 1 ) ).toHaveText(
-			'Sweet paragraph 2'
-		);
-		await expect(
-			editor.canvas
-				.getByRole( 'document', {
-					name: 'Block: Group',
-				} )
-				.getByRole( 'document', {
-					name: 'Block: Title',
-				} )
-		).toBeVisible();
+	// Site Title and Page entities should have been modified.
+	await page.getByRole( 'button', { name: 'Save', exact: true } ).click();
+	await expect(
+		page.locator(
+			'role=region[name="Save panel"] >> role=checkbox[name="Title"]'
+		)
+	).toBeVisible();
+	await expect(
+		page.locator(
+			'role=region[name="Save panel"] >> role=checkbox[name="Test Page"]'
+		)
+	).toBeVisible();
+	await page.getByRole( 'button', { name: 'Cancel', exact: true } ).click();
+}
 
-		// Remove focus from templateOptionsButton button.
-		await editor.canvas.locator( 'body' ).click();
+test.describe( 'Pages', () => {
+	test.beforeAll( async ( { requestUtils } ) => {
+		await requestUtils.activateTheme( 'emptytheme' );
+		await Promise.all( [
+			requestUtils.deleteAllTemplates( 'wp_template' ),
+			requestUtils.deleteAllPages(),
+		] );
+	} );
 
-		// Toggle template preview to "on".
-		await templateOptionsButton.click();
-		await templatePreviewButton.click();
-		await expect( templatePreviewButton ).toHaveAttribute(
-			'aria-pressed',
-			'true'
-		);
+	test.afterAll( async ( { requestUtils } ) => {
+		await requestUtils.activateTheme( 'twentytwentyone' );
+		await Promise.all( [
+			requestUtils.deleteAllTemplates( 'wp_template' ),
+			requestUtils.deleteAllPages(),
+		] );
+	} );
 
-		// Header template area and page content are once again visible.
-		await expect(
-			editor.canvas.getByRole( 'document', {
-				name: 'Block: header',
-			} )
-		).toBeVisible();
-		await expect( paragraphs.nth( 0 ) ).toBeVisible();
-		await expect( paragraphs.nth( 1 ) ).toBeVisible();
-		await expect(
-			editor.canvas.getByRole( 'document', {
-				name: 'Block: Title',
-			} )
-		).toBeVisible();
+	test.beforeEach( async ( { requestUtils, admin } ) => {
+		await Promise.all( [
+			requestUtils.deleteAllTemplates( 'wp_template' ),
+			requestUtils.deleteAllPages(),
+		] );
+		await admin.visitSiteEditor();
+	} );
+
+	test( 'create a new page, edit template and toggle page template preview', async ( {
+		page,
+		editor,
+	} ) => {
+		// Set up
+		await draftNewPage( page );
+		await addPageContent( editor, page );
+
+		// Run assertions
+		await testCreatePage( editor, page );
+		await testToggleTemplatePreview( editor, page );
 	} );
```

Comment:
I love the approach here of deferring to other functions for each part of the e2e test so that the structure is nice and readable. Unfortunately it means the `playwright/expect-expect` lint rule is complaining that there is no assertion in this test. Would it be worth disabling that rule for this test (or this file)? I see `/* eslint-disable playwright/no-conditional-in-test, playwright/expect-expect */` is in use in a few test files.

I wonder if @kevin940726 has any opinions on this one?

Reply:
Personally, I like inlining test commands directly inside the `test` callback so that I don't have to jump around the cursor when reading the test. A little of duplication is a good thing in tests too. By placing assertions outside of the `test` callback is often a bad idea because that messes up the trace viewer when debugging a bit.

In terms of code relocation and naming, I would keep it simple and just inline them in tests with comments. In terms of deduplication and abstraction, I would abstract the "setup" part of the test instead and duplicate the "assertions" part of the test in each test.

That said, these are just my preference and they don't represent the best practices whatsoever 😆. i.e. I won't block this PR because of this suggestion. 🙏

Reply:
>  A little of duplication is a good thing in tests too. By placing assertions outside of the test callback is often a bad idea because that messes up the trace viewer when debugging a bit. .... I would abstract the "setup" part of the test instead and duplicate the "assertions" part of the test in each test.

This makes sense to me. I can relocate the assertions so that they're inside the `test`. It doesn't affect much, just moving the contents of the new test functions back into the `test()` body.

Thanks for the tips 🙇🏻

Reply:
+1 thanks for the guidance here @kevin940726, that all sounds reasonable to me, too!